[TEST] test query and search source parsing from different XContent types

javanna opened the pull request:
We used to test only json parsing as we relied on QueryBuilder#toString which uses the json format. This commit makes sure that we now output the randomly generated queries using a random format, and that we are always able to parse them correctly.

This revealed a couple of issues with binary objects that haven't been migrated yet to be structured Writeable objects. We used to keep them in the format they were sent while parsing, which led to problems when printing them out as we expected them to always be in json format. Also we can't compare different BytesReference objects that hold the same content but in different formats (unless we want to parse them as part of equal and hashcode, doesn't seem like a good idea)  and verify that we have parsed the right objects if they can be different formats. The fix is to always keep binary objects in json format. Best fix would be not to have binary objects, which we'll get to once we are done with the search refactoring.

javanna commented:
@colings86 @cbuescher can you have a look please?

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/TemplateQueryBuilderTests.java:
```diff
@@ -80,7 +80,8 @@ protected void assertBoost(TemplateQueryBuilder queryBuilder, Query query) throw
     public void testUnknownField() throws IOException {
         TemplateQueryBuilder testQuery = createTestQueryBuilder();
-        String queryAsString = testQuery.toString().replace("inline", "bogusField");
+        String testQueryAsString = toXContent(testQuery, randomFrom(XContentType.JSON, XContentType.YAML)).string();
```
Heads up when you merge with master, I just merged another test where the original test query is modified assuming it is json, so that will need the same treatment as you do here I thinkg: https://github.com/elastic/elasticsearch/pull/14255/files#diff-9dc314365d49d84bff0645c2f9dfd7adR356 (and Overwrites in HasChild/HasParentQueryBuilderTests)

javanna replied:
ok will have a look thanks

cbuescher replied:
Just checking AbstractQueryBuilder#toString() which is implemented in ToXContentToBytes. Seems to always use json builder, so I'm not sure if for this kind of test (and the ones mentioned) we need aditional Yaml variant, but its okay to add it I guess.

javanna replied:
yes we always print the query as a string using json. That said toXContent supports different formats, as well as elasticsearch. I think it is good testing with different formats. It would be a bit more complicated in the base class, cause string replacement would need to be different depending on whether we are using yaml or json, but this one was easy. Note that when we do string replacement the binary formats are not used...

cbuescher commented:
LGTM, added a minor comment.

javanna commented:
I have rebased and updated, will wait for @colings86 to have a look though just to make sure that he agrees on the changes made to `SearchSourceBuilder`

colings86 commented:
LGTM
